db/hints/manager.cc

```diff
@@ -849,6 +850,8 @@ future<> manager::end_point_hints_manager::sender::send_one_hint(lw_shared_ptr<s
             }
             return make_ready_future<>();
         }).finally([units = std::move(units), ctx_ptr] {});
+
+        return utils::get_local_injector().inject("hinted_handoff_slow_hint_replay", std::chrono::seconds(1));
```

Comment:
This seems wrong: weren't you supposed to add this line in the body of `finally` lambda?

Reply:
No, I think it is in the right place. The purpose of this sleep is to make sure that we initiate sending of a hint 1 second after previous hint sending was initiated.

The `send_one_hint` function works as follows:

1. Wait for send units (line 814/815)
2. Spawn discarded future which sends the hint and holds the send units (line 816/817)
3. Sleep 1 second if error injection is enabled (line 854)

Next invocation of `send_one_hint` does not start until the previous one has finished, so this limits hint replay to 1 hint per second.

----

Notice that the `finally` lambda is attached to the discarded future. If I put the error injection sleep there, then it would just cause the discarded future to resolve one second later, not `send_one_hint`.

If, additionally, I made sure that send units are released after the error injection sleep, then the following would happen:

1. `send_one_hint` executes 128 times (current hints concurrency limit on one shard) and spawns futures which send hints.
2. The next invocation of `send_one_hint` needs to wait because there are no send units available - hint-sending futures will release them after 1 second.
3. The hint-sending futures release their send units roughly at the same time and we go back to step 1.

So, instead of limiting hint replay to 1 hint/s, we would limit hint replay to `max concurrency on shard` hints/s.

Reply:
Right. I missed that this `finally` relates to the `with_gate()`.